core/src/main/java/com/linecorp/armeria/common/Flags.java

```diff
@@ -520,18 +525,32 @@ public static TransportType transportType() {
      *
      * <p>This flag is enabled by default for supported platforms. Specify the
      * {@code -Dcom.linecorp.armeria.useOpenSsl=false} JVM option to disable it.
+     *
+     * @deprecated Use {@link #tlsEngineType()} and {@code -Dcom.linecorp.armeria.tlsEngineType=openssl}.
      */
+    @Deprecated
     public static boolean useOpenSsl() {
-        if (useOpenSsl != null) {
-            return useOpenSsl;
+        return tlsEngineType() == TlsEngineType.OPENSSL;
+    }
+
+    /**
+     * Returns the {@link TlsEngineType} that will be used for processing TLS connections.
+     *
+     * <p>The default value of this flag is "openssl", which means the default {@link TlsEngineType} will
+     * be used. Specify the {@code -Dcom.linecorp.armeria.tlsEngineType=<jdk|openssl>} JVM option to override
+     * the default.</p>
+     */
+    public static TlsEngineType tlsEngineType() {
+        if (TLS_ENGINE_TYPE != null) {
+            return TLS_ENGINE_TYPE;
         }
         setUseOpenSslAndDumpOpenSslInfo();
-        return useOpenSsl;
+        return TLS_ENGINE_TYPE;
     }
 
     private static void setUseOpenSslAndDumpOpenSslInfo() {
-        final boolean useOpenSsl = getValue(FlagsProvider::useOpenSsl, "useOpenSsl");
-        if (!useOpenSsl) {
+        TLS_ENGINE_TYPE = getValue(FlagsProvider::tlsEngineType, "tlsEngineType");
```

Comment:
For smooth migration, `-Dcom.linecorp.armeria.useOpenSsl=false` needs to return `TlsEngineType.JDK` if no `TlsEngineType` is specified by a user. Otherwise, an OpenSSL engine is used unexpectedly.

How about adding `getUserValue()` method that ignores `DefaultFlagsProvider`'s values and returns only user-defined values?
```java
final Boolean useOpenSsl = getUserValue(FlagsProvider::useOpenSsl, "useOpenSsl");
final TlsEngineType tlsEngineType = getUserValue(FlagsProvider::tlsEngineType, "tlsEngineType");

if (useOpenSsl == null) {
   // Respect tlsEngineType
} else if (tlsEngineType == null) {
  // Respect useOpenSsl
} else {
   // Check if the values of useOpenSsl and tlsEngineType are compatible. 
}
```

Reply:
I've added `getUserValue()` which ignores `DefaultFlagsProvider`. 

In the case of incompatible `useOpenSsl` and `tlsEngineType`, I left a warning log and used the value from `tlsEngineType`. I chose `tlsEngineType` instead of `useOpenSsl` because in that case, user is aware of and using `-Dcom.linecorp.armeria.tlsEngineType`, so I think it's might be fine to use the value of `tlsEngineType`. 
```java
else {
      if (useOpenSsl != (tlsEngineType == TlsEngineType.OPENSSL)) {
          logger.warn("useOpenSsl({}) and tlsEngineType({}) are incompatible, tlsEngineType will be used",
                      useOpenSsl, tlsEngineType);
      }
      TLS_ENGINE_TYPE = tlsEngineType;
}
```